File: plaoc/src/client/util/rect.ts
export type $Rect = {
  x: number;
  y: number;
  width: number;
  height: number;
};
export const rectToDom = (rect: $Rect) =>
  new DOMRect(rect.x, rect.y, rect.width, rect.height);
export const domRectToJson = (domRect: DOMRect): $Rect => ({
  x: domRect.x,
  y: domRect.y,
  width: domRect.width,
  height: domRect.height,
});

File: next/kmp/window/src/commonMain/kotlin/org/dweb_browser/sys/window/core/Rect.kt
package org.dweb_browser.sys.window.core

import kotlinx.serialization.Serializable

/**
 * 窗口大小与位置
 *
 * 默认值是NaN，这种情况下，窗口构建者需要自己对其进行赋值
 */
@Serializable
data class Rect(
  val x: Float = Float.NaN,
  val y: Float = Float.NaN,
  val width: Float = Float.NaN,
  val height: Float = Float.NaN,
) {
  companion object {
    val Zero = Rect(0f, 0f, 0f, 0f)
  }

  fun toMutable() = Mutable(x, y, width, height)
  class Mutable(
    var x: Float,
    var y: Float,
    var width: Float,
    var height: Float,
  ) {
    fun toImmutable() = Rect(x, y, width, height)
  }
}